src/microbe_stage/systems/MucocystSystem.cs

```diff
@@ -0,0 +1,59 @@
+﻿namespace Systems;
+
+using System;
+using Components;
+using DefaultEcs;
+using DefaultEcs.System;
+using DefaultEcs.Threading;
+
+/// <summary>
+///   Handles gaining immortality from mucocyst ability
+/// </summary>
+[With(typeof(MicrobeControl))]
+[With(typeof(Health))]
+[With(typeof(CompoundStorage))]
+[Without(typeof(AttachedToEntity))]
+[WritesToComponent(typeof(CellProperties))]
+[RunsAfter(typeof(OrganelleComponentFetchSystem))]
+[RunsBefore(typeof(MicrobeMovementSystem))]
+[RuntimeCost(7)]
+public sealed class MucocystSystem : AEntitySetSystem<float>
+{
+    private Compound mucilageCompound = SimulationParameters.Instance.GetCompound("mucilage");
+
+    public MucocystSystem(World world, IParallelRunner runner) :
+        base(world, runner)
+    {
+    }
+
+    protected override void Update(float delta, in Entity entity)
+    {
+        // Handles invulnerability from mucocyst. Other buffs/debuffs from mucocyst are in related systems
+        ref var control = ref entity.Get<MicrobeControl>();
+
+        if (control.State == MicrobeState.MucocystShield)
+        {
+            entity.Get<Health>().Invulnerable = true;
+            entity.Get<CompoundStorage>().Compounds.
+                Compounds[mucilageCompound] -= Math.Min(
+                Constants.MUCOCYST_MUCILAGE_DRAIN * delta,
+                entity.Get<CompoundStorage>().Compounds.Compounds[mucilageCompound]);
+
+            if (entity.Get<CompoundStorage>().Compounds.Compounds[mucilageCompound] <= 0)
+                control.State = MicrobeState.Normal;
+
+            var membrane = entity.Get<CellProperties>().CreatedMembrane;
+
+            if (membrane != null)
+                membrane.SetMucocystEffectVisible(true);
+        }
+        else
+        {
+            entity.Get<Health>().Invulnerable = CheatManager.GodMode;
```

Comment:
Doesn't this set AI cells to be invulnerable as well?

Also it is pretty non-optimal to constantly have to call methods on the membrane. Most systems are designed so that they need to at most like check a single bool flag to know if they need to process an entity or not.

There isn't a clear way to do that here, but it might be more performant to keep an internal list of active mucocyst effects in this system and then turning off the effects for cells that don't mark themselves as being still active in the update method.

Reply:
This was also not addressed yet.